### PR TITLE
[serve] Get new handle to controller if killed

### PR DESCRIPTION
**Pull request by shrekris-anyscale**
## Why are these changes needed?

`serve shutdown` is not idempotent with the new Serve CLI. When serve shuts down, it kills the controller. The REST API does not refresh its cached controller handle, so it attempts to make requests to a dead actor, which fail.

This change updates the REST API and `serve.start()` to refresh the controller handle if the controller has been killed.

## Related issue number

N/A

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
     - New unit tests are added to `test_cli.py` and `test_standalone2.py`.

### CODE REVIEW

File: python/ray/serve/api.py [hunk 1]
```diff
@@ -108,9 +109,27 @@ def _get_controller_namespace(
...
+            to the controller.
+    """
+
+    try:
+        if _global_client is not None:
+            if _health_check_controller:
+                ray.get(_global_client._controller.check_alive.remote())
+            return _global_client
+    except RayActorError:
+        logger.debug("The cached controller has died. Reconnecting.")
```
Comment (edoakes) [inline python/ray/serve/api.py:132]: we should clear `_global_client` in this case

File: dashboard/optional_utils.py [hunk 2]
```diff
@@ -268,8 +267,6 @@ async def decorator(self, *args, **kwargs):
                         ray.shutdown()
                         raise e from None
 
-                if connect_to_serve:
-                    serve.start(detached=True, _override_controller_namespace="serve")
```
Comment (edoakes) [inline dashboard/optional_utils.py:272]: why move out of decorator?

Comment (shrekris-anyscale) [reply dashboard/optional_utils.py]: Resolved offline; I thought the decorator ran only once per function, but it runs once per call. I moved `serve.start()` back into the decorator.

File: python/ray/serve/api.py [hunk 3]
```diff
@@ -108,9 +109,28 @@ def _get_controller_namespace(
...
```
Comment (edoakes) [inline python/ray/serve/api.py:132]: can you make this an info level log? might be useful to see in logs in some scenarios

**Review by edoakes — APPROVED** [review]
one more nit

**Pull request merged**